core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -0,0 +1,170 @@
+    }
+
+    new TransactionMarkerChannelManager(config,
+      metadataCache,
+      txnMarkerPurgatory,
+      sendThread,
+      channel)
+  }
+
+
+  private[transaction] def requestGenerator(transactionMarkerChannel: TransactionMarkerChannel,
+                                            txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]): () => Iterable[RequestAndCompletionHandler] = {
+    def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+      transactionMarkerChannel.brokerStateMap.flatMap {case (brokerId: Int, destAndMarkerQueue: DestinationBrokerAndQueuedMarkers) =>
+        val markersToSend: java.util.List[CoordinatorEpochAndMarkers] = new util.ArrayList[CoordinatorEpochAndMarkers] ()
+        destAndMarkerQueue.markersQueue.drainTo (markersToSend)
+        markersToSend.groupBy{ epochAndMarker => (epochAndMarker.metadataPartition, epochAndMarker.coordinatorEpoch) }
+          .map { case((metadataPartition: Int, coordinatorEpoch:Int), buffer: mutable.Buffer[CoordinatorEpochAndMarkers]) =>
+            val txnMarkerEntries = buffer.flatMap { x => x.txnMarkerEntry }.asJava
+            val requestCompletionHandler = new TransactionMarkerRequestCompletionHandler(
+              transactionMarkerChannel,
+              txnMarkerPurgatory,
+              CoordinatorEpochAndMarkers(metadataPartition, coordinatorEpoch, txnMarkerEntries),
+              brokerId)
+            RequestAndCompletionHandler(destAndMarkerQueue.destBrokerNode, new WriteTxnMarkersRequest.Builder(coordinatorEpoch, txnMarkerEntries), requestCompletionHandler)
+          }
+      }
+    }
+    generateRequests
+  }
+}
+
+
+
+class TransactionMarkerChannelManager(config: KafkaConfig,
+                                      metadataCache: MetadataCache,
+                                      txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
+                                      interBrokerSendThread: InterBrokerSendThread,
+                                      transactionMarkerChannel: TransactionMarkerChannel) extends Logging {
+
+  type WriteTxnMarkerCallback = Errors => Unit
+
+  def start(): Unit = {
+    interBrokerSendThread.start()
+  }
+
+  def shutdown(): Unit = {
+    interBrokerSendThread.shutdown()
+    transactionMarkerChannel.clear()
+  }
+
+
+  def addTxnMarkerRequest(metadataPartition: Int, metadata: TransactionMetadata, coordinatorEpoch: Int, completionCallback: WriteTxnMarkerCallback): Unit = {
+    val metadataToWrite = metadata synchronized metadata.copy()
+
+    if (!transactionMarkerChannel.maybeAddPendingRequest(metadataPartition, metadata))
+      // TODO: Not sure this is the correct response here?
+      completionCallback(Errors.INVALID_TXN_STATE)
+    else {
+      val delayedTxnMarker = new DelayedTxnMarker(metadataToWrite, completionCallback)
+      txnMarkerPurgatory.tryCompleteElseWatch(delayedTxnMarker, Seq(metadata.pid))
```

Comment:
Since the timeout in DelayedTxnMarker is infinite. I am wondering if we really need a txnMarkerPurgatory. In TransactionMarkerRequestCompletionHandler, we are already updating the pending partitions as the client response comes back. The response that removes the last pending partition can just trigger the calling of completionCallback.